src/Rector/v12/v0/typo3/ReplaceExpressionBuilderMethodsRector.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Ssch\TYPO3Rector\Rector\v12\v0\typo3;
+
+use PhpParser\Node;
+use PhpParser\Node\Expr\MethodCall;
+use PHPStan\Type\ObjectType;
+use Rector\Core\Rector\AbstractRector;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @changelog https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/12.0/Deprecation-97354-ExpressionBuilderMethodsAndXAndOrX.html
+ * @see \Ssch\TYPO3Rector\Tests\Rector\v12\v0\typo3\ReplaceExpressionBuilderMethodsRector\ReplaceExpressionBuilderMethodsRectorTest
+ */
+final class ReplaceExpressionBuilderMethodsRector extends AbstractRector
+{
+    /**
+     * @return array<class-string<Node>>
+     */
+    public function getNodeTypes(): array
+    {
+        return [MethodCall::class];
+    }
+
+    /**
+     * @param MethodCall $node
+     */
+    public function refactor(Node $node): ?Node
+    {
+        if ($this->shouldSkip($node)) {
+            return null;
+        }
+
+        if (! $this->isNames($node->name, ['andX', 'orX'])) {
+            return null;
+        }
+
+        $methodName = $this->isName($node->name, 'andX') ? 'and' : 'or';
+
+        return $this->nodeFactory->createMethodCall($node->var, $methodName, $node->args);
+    }
+
+    /**
+     * @codeCoverageIgnore
+     */
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition('Replaces ExpressionBuilder methods orX() & andX()', [new CodeSample(
+            <<<'CODE_SAMPLE'
+$rows = $queryBuilder
+  ->select(...)
+  ->from(...)
+  ->where(
+    $queryBuilder->expr()->andX(...),   // replace with and(...)
+    $queryBuilder->expr()->orX(...)     // replace with or(...)
+  )
+  ->executeQuery()
+  ->fetchAllAssociative();
+CODE_SAMPLE
+            ,
+            <<<'CODE_SAMPLE'
+$rows = $queryBuilder
+  ->select(...)
+  ->from(...)
+  ->where(
+    $queryBuilder->expr()->and(...), // replacement for andX(...)
+    $queryBuilder->expr()->or(...)   // replacement for orX(...)
+  )
+  ->executeQuery()
+  ->fetchAllAssociative();
+CODE_SAMPLE
+        )]);
+    }
+
+    private function shouldSkip(MethodCall $methodCall): bool
+    {
+        return ! $this->isObjectType(
+            $methodCall,
+            new ObjectType('TYPO3\CMS\Core\Database\Query\Expression\ExpressionBuilder')
```

Comment:
In my mind this is the correct ObjectType, but the test says otherwise..
I don't know how to handle this/ how the chaining of methods is working

Reply:
I could adapt one test case, where the ExpressionBuilder is set specifically.
I stepped into the chained methods as well but could not really figure out what's happening there.
On comparison with other rules the chaining is handled by checking for the initial ObjectType, e.g. here: \Ssch\TYPO3Rector\Rector\v11\v0\RemoveLanguageModeMethodsFromTypo3QuerySettingsRector

But seems to be different here.

Reply:
So i changed the stubs and now it works?
I'm a little irritated tbh, but the tests are obviously green and the shouldskip method really only checks for the ExpressionBuilder, so i would guess this is a proper and fully working rule without side effects.
Just wondering what was not fitting before that..